### DISCUSSION TITLE
CatchAll with possibility to validate keys?

### DISCUSSION THREAD

**Discussion by reuzel** [colinhacks/zod#1925] [Q&A]
Hi,

I'm trying to validate an object structure, where some (but not all!) fields are multi-lingual. The field names are then defined based on a pattern: `${field}#${locale}`. For example, the following should pass:

```json
{
    "id": "123"
    "client_name": "my great app",
    "client_name#nl_NL": "Mijn geweldige applicatie",
}
```
As there are so many locales, it is kind of impossible to mention all versions in an object schema. Instead, I was looking for something like catchAll, but then one that does not only allow me to verify the values of the not-specifically mentioned fields, but also the names of the fields. That would allow me to do something like:

```typescript
 const schema = z.object({
   id: z.string(),
   client_name: z.string(),
 })
.catchAll(z.string().refine(key=> /** custom key validation logic here **/), z.string())
.requires({id:true});
```

Any suggestion on how to do this?

Regards,
Joost

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const schema = z.object( {
    id: z.string(),
    client_name: z.record( z.string() ).refine( value =>
        Object.keys( value ).find( key => key.split( '_' ).length === 2 )
        // for each key: checks that there is an
        // underscore with a string on each side
        // or do something else
    ),
} )

console.log(
    schema.safeParse( {
        id: '123',
        client_name: {
            en_US: 'my great app',
            nl_NL: 'Mijn geweldige applicatie',
        }
    } ).success
) // true
```

**Reply by joostreuzel** [reply]
Not exactly, the locale is embedded in the key of the top-object. Regrettably I'm  not in the position to change that as it is part of a standard (see: section 2.1 and example in section 3.1 of [OIDC Client Registration Spec](https://openid.net/specs/openid-connect-registration-1_0.html#LanguagesAndScripts)).

What I've done now is this:

```typescript
function multiLangObject<T extends { [key: string]: zod.Schema }>(
    definition: T
): zod.ZodRecord<
    zod.ZodEffects<zod.ZodString, string, string>,
    zod.ZodEffects<zod.ZodAny, unknown, unknown>
> {
    const keys = Object.getOwnPropertyNames(definition);

    let key: string | undefined = undefined;

    const keySchema = zod.string().refine((val) => {
        const parts = val.split('#', 2);
        const ok = keys.includes(parts[0]) && (parts[1] === undefined || validateLocale(parts[1]));
        key = ok ? parts[0] : undefined;
        return ok;
    });

    const valueSchema = zod.any().refine((val) => {
        if (!key) return false;
        const result = definition[key].safeParse(val);
        return result.success;
    });

    return zod.record(keySchema, valueSchema);
}
```
This snippet uses the key validation possibilities of a record to check if a key is listed and if the key has an optional locale. The bad thing about this implementation is that the key name is 'remembered' to make sure the right value validation is executed. This assumes that record values are always processed right after processing the corresponding key, which is a big assumption but seems to work.

and I use this to "and" the record to my object definition:

```typescript

const schema = zod.object({
   id: zod.string(),
   //omitted for brevity
}).and(multiLangObject({
  client_name: zod.string(),
  tos_uri: zod.string().url(),
});
```

this allows me to process something like:

```typescript
{
    id: "123",
    client_name: "my great app",
    "client_name#nl_NL": "mijn geweldige applicatie"
}
```

While working on this I found that it would be great if in object definitions a template or regex could be provided as key, such that all object keys that match the template/regex would be validated against the schema associated with it...

Regards,
Joost

**Reply by joostreuzel** [reply]
Actually my approach doesn't work nicely either. Doing an 'and' between an object and record basically kills all typing...

I think I will move towards preprocessing, filtering out all localized fields and put in an object as described by you above...

**Reply by JacobWeisenburger** [reply]
great. Does that mean that my answer is satisfactory?

**Reply by reuzel** [reply]
Yes and no. 

Yes: with some preprocessing of the data I got it working... 

No: More elegant would be the opportunity to use some template in object definitions to match keys to. Especially if parameters from that template would be available in type checking actions (such as in superRefine context). That way, I could have achieved this without preprocessing. For example it would be great if something like this could be written...

```javascript
const schema = z.object({
   id: z.string(),
   `name#{locale}`: z.string().superRefine((val, ctx) =>{ return isValidLocale(ctx.key.params['locale'])}),
});
```

This would succeed:
```javascript
schema.parse({
   id: "123",
   "name#nl_NL": "naam",
   "name#fr_FR": "nom",
})
```

But this wouldn't (as not-a-locale isn't conforming to the locale syntax standards):
```javascript
schema.parse({
   id: "123",
   "name#not-a-locale": "naam",
})
```

See it as a feature request...
